### ISSUE TITLE
assert_throw returns confusing message if ArgumentError is thrown

### ISSUE THREAD

**Issue by chrisnicola** [seattlerb/minitest#622]
If a normal ArgumentError is raised during a test assert_throws will catch it and a very confusing message is returned because of this code:

```
    rescue ArgumentError => e     # 1.9 exception
      default += ", not #{e.message.split(/ /).last}"
```

So you get the last part of the argument error returned as the name of a missing argument like:

`Expected :denied to have been thrown, not :id.`

**Comment by zenspider** [comment]
Hopping on a plane. Can you craft a failing test? Crib from the existing throw tests to make it easier.

--  
Ryan Davis
Sent with Airmail

**Comment by chrisnicola** [comment]
Sure here is a trivial example:

```
      assert_throws(:something) {
        method_not_here
        throw :something
      }
```

It seems to be that the purpose behind this code was to catch an unexpected throw. Say I do `assert_throws(:something)` and then I do `throw :something_else`. Which normally raises an exception when uncaught.

In Ruby 2 the exception is `UncaughtThrowException`. Based on the comments I can only assume that for older versions it isn't that specific. I don't know there is definitely some `wtf` in there so...

I would suggest that it makes sense to just let the exception throw. It is unexpected, it isn't as if expected something to be thrown means that something else being thrown isn't in fact an unexpected exception. It probably is.

**Comment by chrisnicola** [comment]
These are all the exceptions being handled here:

```
    rescue ThreadError => e       # wtf?!? 1.8 + threads == suck
      default += ", not \:#{e.message[/uncaught throw \`(\w+?)\'/, 1]}"
    rescue ArgumentError => e     # 1.9 exception
      default += ", not #{e.message.split(/ /).last}"
    rescue NameError => e         # 1.8 exception
      default += ", not #{e.name.inspect}"
```

So basically any method call that throws and ArgumentError or NameError will potentially return a confusing assert failure message.

**Comment by zenspider** [comment]
This seems to have been fixed in #623. Closing.
